src/feat/signal-distort.cc

```diff
@@ -72,5 +72,81 @@ void TimeStretch(const MatrixBase<BaseFloat> &input_egs,
   perturb_egs->CopyFromMat(out_mat);
 }
 
+// This function add the noise to the orginial signal. We should not normalize 
+// the signal level of the orginial signal. According to SNR, we rescale the noise
+// and add it. So that the perturbed signal is created. 
+void PerturbXvectorSignal::ApplyAdditiveNoise(const MatrixBase<BaseFloat> &input_eg,
+                                              const Matrix<BaseFloat> &noise_eg,
+                                              Matrix<BaseFloat> *perturbed_eg) {
+  // In the version, we ask the noise_cols == input_cols.
+  int32 input_rows = input_eg.NumRows(), input_cols = input_eg.NumCols();  
+  KALDI_ASSERT(noise_eg.NumCols() == input_cols);
+
+  // According to the rows of noise_eg, form the noise_mat
+  // repeat the noise_eg blocks to have a new block which is longer than input_eg
+ 
+  // As the noise_eg is very huge and the input_eg is small normally,
+  // so we'd better not reload the "noise_eg" matrix
+  // select the noise range
+
+  Matrix<BaseFloat> selected_noise_mat;
+  selected_noise_mat.Resize(input_rows, input_cols);
+  
+  int32 noise_rows = noise_eg.NumRows();
+  int32 start_row_ind = RandInt(0, noise_rows - input_rows);
+  
+  if (noise_eg.NumRows() < input_rows) {
+    int32 indices[input_rows];
+    for (int32 i=0; i < input_rows; ++i) {
+      indices[i] = (start_row_ind + i) % noise_eg.NumRows();
+    }
+    selected_noise_mat.CopyRows(noise_eg, indices);
+  } else {
+    selected_noise_mat.AddMat(1.0, noise_eg.Range(start_row_ind, input_rows,
+                                                  0, input_cols));
+  }
+
+  // compute the energy of noise and input
+  Matrix<BaseFloat> input_energy_mat(input_rows, input_cols);
+  input_energy_mat.AddMatMatElements(1.0, input_eg, input_eg, 0.0);
+  double input_energy = input_energy_mat.Sum();
+  Matrix<BaseFloat> noise_energy_mat(input_rows, input_cols);
+  noise_energy_mat.AddMatMatElements(1.0, selected_noise_mat, selected_noise_mat, 0.0);
+  double noise_energy = noise_energy_mat.Sum();
+
+  // In Energy domain, SNR=20log10(S/N). 
+  // 10^(SNR/20) = input_energy / (scale^2 * noise_energy)
+  double scale = input_energy / noise_energy / (pow(10,opts_.snr/20));
+  scale = sqrt(scale);
+  
+  // Add noise mat to input_eg mat
+  perturbed_eg->Resize(input_rows, input_cols);
+  perturbed_eg->CopyFromMat(input_eg);
+  perturbed_eg->AddMat(scale, selected_noise_mat);
+}
+
+void PerturbXvectorSignal::ApplyDistortion(const MatrixBase<BaseFloat> &input_egs,
+                                           Matrix<BaseFloat> *perturb_egs) {
+    // conduct ApplyAdditiveNoise
+  if (!opts_.add_noise_rspecifier.empty()) {
```

Comment:
I think the best strategy is to have a add-noise option  in PertubXvectorOption as noise rspecifier not noise examples.
--add-noise=noise.scp, where noise.scp corresponds to features for different noises. You can randomly select different noises.
Then you no longer need to pass noise matrix to PerturbExample and you can easily pass noise rspecifier using --add-noise option.
You don't need to change ApplyAdditiveNoise class. You just need to check if add-noise is not empty  in ApplyDistortion and the read matrix of noise using BaseFloatMatrixReader and pass it to ApplyAdditiveNoise.